ui/app/adapters/watchable.js

```diff
@@ -0,0 +1,144 @@
+    }
+
+    return this.ajax(url, 'GET', {
+      data: params,
+    }).catch(error => {
+      if (error instanceof AbortError) {
+        return [];
+      }
+      throw error;
+    });
+  },
+
+  findRecord(store, type, id, snapshot, additionalParams = {}) {
+    let [url, params] = this.buildURL(type.modelName, id, snapshot, 'findRecord').split('?');
+    params = assign(queryString.parse(params) || {}, this.buildQuery(), additionalParams);
+
+    if (get(snapshot || {}, 'adapterOptions.watch')) {
+      params.index = this.get('watchList').getIndexFor(url);
+    }
+
+    return this.ajax(url, 'GET', {
+      data: params,
+    }).catch(error => {
+      if (error instanceof AbortError) {
+        return {};
+      }
+      throw error;
+    });
+  },
+
+  reloadRelationship(model, relationshipName, watch = false) {
+    const relationship = model.relationshipFor(relationshipName);
+    if (relationship.kind !== 'belongsTo' && relationship.kind !== 'hasMany') {
+      throw new Error(
+        `${relationship.key} must be a belongsTo or hasMany, instead it was ${relationship.kind}`
+      );
+    } else {
+      const url = model[relationship.kind](relationship.key).link();
+      let params = {};
+
+      if (watch) {
+        params.index = this.get('watchList').getIndexFor(url);
+      }
+
+      if (url.includes('?')) {
+        params = assign(queryString.parse(url.split('?')[1]), params);
+      }
+
+      return this.ajax(url, 'GET', {
+        data: params,
+      }).then(
+        json => {
+          this.get('store').pushPayload(relationship.type, {
+            [relationship.type]: makeArray(json),
+          });
+        },
+        error => {
+          if (error instanceof AbortError) {
+            return relationship.kind === 'belongsTo' ? {} : [];
+          }
+          throw error;
+        }
+      );
+    }
+  },
+
+  handleResponse(status, headers, payload, requestData) {
+    const newIndex = headers['x-nomad-index'];
+    if (newIndex) {
+      this.get('watchList').setIndexFor(requestData.url, newIndex);
+    }
+
+    return this._super(...arguments);
+  },
+
+  cancelFindRecord(modelName, id) {
+    const url = this.urlForFindRecord(id, modelName);
+    const xhr = this.get('xhrs')[url];
+    if (xhr) {
+      xhr.abort();
+    }
+  },
+
+  cancelFindAll(modelName) {
+    const xhr = this.get('xhrs')[this.urlForFindAll(modelName)];
+    if (xhr) {
+      xhr.abort();
+    }
+  },
+
+  cancelReloadRelationship(model, relationshipName) {
+    const relationship = model.relationshipFor(relationshipName);
+    if (relationship.kind !== 'belongsTo' && relationship.kind !== 'hasMany') {
+      throw new Error(
+        `${relationship.key} must be a belongsTo or hasMany, instead it was ${relationship.kind}`
+      );
+    } else {
+      const url = model[relationship.kind](relationship.key).link();
+      const xhr = this.get('xhrs')[url];
+      if (xhr) {
+        xhr.abort();
+      }
+    }
+  },
```

Comment:
👏 👏 :clap: wow this file - super bummer there's not a better way to cancel requests than to manually track and reach down into the jquery XHR object, but still this is 👌

Reply:
Yeah. Though with a promise-based workflow, there isn't a good place to expose the XHR. And promises aren't cancelable, so that avenue doesn't work either. Maybe we need `AjaxPromiseProxies` to come back from store methods.